### PR TITLE
Utility tweaks for Memory Constrained

### PR DESCRIPTION
**Pull request by ludziej**
# Description
[Neptune](https://app.neptune.ai/o/pmtest/org/llm-random/runs/details?viewId=99f79ace-8d2e-4454-824e-1d95bd4e481d&detailsTab=metadata&shortId=LLMRANDOM-12257&type=run&path=loss_interval%2F&attribute=10_tokens_per_active) link for active plot

[Neptune](https://app.neptune.ai/o/pmtest/org/llm-random/runs/details?viewId=9cb5eb14-505d-4f9a-88b0-e55f2fc0bd4e&detailsTab=metadata&shortId=LLMRANDOM-12288&path=&attribute=lr) link for scheduler params.

Add loss plots, measured in tokens per active parameter.

Removes zip strict=True function that made the repository not compatible with python 3.9

Adds default final_lr_step=steps for -1 and warmup as a float (of n_steps)
Adds config for memory constrained scaling laws

# Checklist
   - [ ] Is this PR focused on a single feature without the possibility of being divided into smaller PRs?
   - [ ] Are all variables set with logical defaults that are backward compatible?
   - [ ] I attached link to neptune if it was necessary
   - [ ] Have you successfully executed the linter and tests?
   - [ ] Are all functions concise and don't require splitting?
   - [ ] Is there documentation for arguments and complex logic?
   - [ ] Are there no temporary solutions in the code? Should a task be added to Trello to refine the code?
   - [ ] Are all functions placed in appropriate files?
   - [ ] Is the PR name meaningful and descriptive?
   - [ ] Is PR description provided, or unnecessary?
  
Other questions:
   - [ ] If I made changes to the requirements or anything related to the Singularity image I did generate new image and upload a new image to the clusters
   - [ ] If this change is significant enough I notified all individuals working with this repository (@channel in slack channel)
   - [ ] If there is a need for a second reviewer I requested additional review 

# Reviewer's checklist:
   - [ ] Does the Neptune link work and does it actually compare the situation before and after the PR?
   - [ ] Do I understand the code?
   - [ ] Do I think another reviewer is not needed or I commented that this PR need another reviewer?

### CODE REVIEW

File: lizrd/support/misc.py [hunk 1]
```diff
@@ -25,6 +25,18 @@ def count_parameters(model, args, VOCAB_SIZE):
     return model_n_params
 
 
+def count_moe_non_emb_active_params(args):
+    return (
+        args.dmodel**2
+        * (4 + 2 * (args.effective_dff_x if args.effective_dff_x is not None else 4))
+        * args.n_blocks
+    )
```
Comment (crewtool) [inline lizrd/support/misc.py:33]: Why not dmodel, effective_dff_x, n_blocks instead of args?

File: lizrd/support/misc.py [hunk 2]
```diff
@@ -25,6 +25,18 @@ def count_parameters(model, args, VOCAB_SIZE):
...
+
+
+def count_tokens_per_step(args):
+    return args.batch_size * args.cutoff
```
Comment (crewtool) [inline lizrd/support/misc.py:37]: same as above

File: lizrd/train/scheduler.py [hunk 3]
```diff
@@ -10,15 +10,19 @@ def get_scheduler(
 ) -> "AbstractLRScheduler":
     if ratios_in_group_order is None:
         ratios_in_group_order = [1.0]
+    if args.lr_warmup_percent is not None:
+        assert not args.lr_warmup_steps  # cannot set both
```
Comment (crewtool) [inline lizrd/train/scheduler.py:14]: You should give a message instead of a comment

File: .gitignore [hunk 4]
```diff
@@ -163,4 +163,6 @@ full_config*.yaml
 # yaml / json schemas
 schema.json
 
-wandb
\ No newline at end of file
+wandb
+
+local_search_configs
```
Comment (crewtool) [inline .gitignore:168]: How about local_configs / custom_configs?

File: lizrd/support/misc.py (hunk 1 above)
Comment (crewtool) [reply lizrd/support/misc.py:33]: What if dff is not 4 * dmodel? We should check that.

File: lizrd/support/logging.py [hunk 5]
```diff
@@ -195,6 +221,7 @@ def report_scalar(
         value: float,
         iteration: int,
         series: Optional[str] = None,
+        token_scale: bool = False,
```
Comment (crewtool) [inline lizrd/support/logging.py:224]: Why we can't just log tokens for every run?

File: configs/experiments/constrained_scaling_laws/common.yaml [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+  batch_size: 512
+  cutoff: 512
+  final_lr_step: -1
+  lr_warmup_percent: 0.01
+  lr_warmup_steps: 0
+  scheduler: cosine
+  final_lr_fraction: 0.1
+  init_scale: 0.1
+  learning_rate: 0.0005
+  weight_decay: 0.01
```
Comment (kuba-krj) [inline configs/experiments/constrained_scaling_laws/common.yaml:37]: Let's change to 0.1

File: lizrd/core/initialization.py [hunk 7]
```diff
@@ -2,11 +2,9 @@
 from torch.nn.init import trunc_normal_
 from functools import partial
 import torch.nn as nn
-from typing import Literal, TypeAlias
+from typing import Literal
 
-ValidInitType: TypeAlias = Literal[
-    "kaiming_uniform", "truncated_normal", "truncated_normal_fixed"
-]
+ValidInitType = Literal["kaiming_uniform", "truncated_normal", "truncated_normal_fixed"]
```
[the remark this answers is not in the document]
Comment (ludziej) [reply lizrd/core/initialization.py:7]: for python 3.9 combatility

File: configs/experiments/constrained_scaling_laws/common.yaml [hunk 8]
```diff
@@ -0,0 +1,49 @@
...
+
+  batch_size: 512
+  cutoff: 512
+  final_lr_step: -1
+  lr_warmup_percent: 0.01
+  lr_warmup_steps: 0
+  scheduler: cosine
+  final_lr_fraction: 0.1
+  init_scale: 0.1
+  learning_rate: 0.0005
```
[the remark this answers is not in the document]
Comment (ludziej) [reply configs/experiments/constrained_scaling_laws/common.yaml]: changed

File: research/blanks/trainer.py [hunk 9]
```diff
@@ -298,6 +298,7 @@ def _log_train_stats(self, loss_value, step, aux_info):
                     title=name,
                     value=stats.acc / stats.interval,
                     iteration=step,
+                    token_scale=True,
```
[the remark this answers is not in the document]
Comment (ludziej) [reply research/blanks/trainer.py:301]: ok, removed it

File: lizrd/train/scheduler.py (hunk 3 above)
Comment (ludziej) [reply lizrd/train/scheduler.py:14]: ok, gonna give an exception

File: .gitignore (hunk 4 above)
Comment (ludziej) [reply .gitignore:168]: this is a folder that already exists from previous PR, for configs from local search, I think the name is descriptive enough

File: lizrd/support/logging.py (hunk 5 above)
Comment (ludziej) [reply lizrd/support/logging.py:224]: Ok, I'm gonna make this default

File: lizrd/train/scheduler.py (hunk 3 above)
Comment (crewtool) [reply lizrd/train/scheduler.py:14]: I meant somthing like this
```suggestion
        assert not args.lr_warmup_steps, "You cannot set both lr_warmap_percent and lr_warmap_steps"
```
alternatively such assert could be made using argparse lib

File: .gitignore (hunk 4 above)
Comment (crewtool) [reply .gitignore:168]: yup, at first i thought it is for your local configs, but then relized it is for local_search :D

File: lizrd/support/logging.py (hunk 5 above)
Comment (ludziej) [reply lizrd/support/logging.py:224]: Ok sorry, I this should be false, because by default we don't want ANY metric to be in token scale, only loss dependent ones, which are directly specified.

File: lizrd/support/misc.py (hunk 1 above)
Comment (ludziej) [reply lizrd/support/misc.py:33]: Ok yup, I'll change to args.dff if args.effective_dff_x is none

Comment (ludziej) [reply lizrd/support/misc.py:33]: I really think passing these arguments here would add unnecessary, error-prone verbosity to function calls. This function is meant to use any parameters that are needed to calculate activated params, the caller does not have to know what's used under the hood.

**Comment by ludziej** [comment]
I've addressed all the comments & applied requested changes

File: lizrd/support/logging.py [hunk 10]
```diff
@@ -120,12 +128,30 @@ def get_metric_with_flop_scale(self, value: float, iteration: int):
...
+        tokens_passed = iteration * self.args.get("tokens_per_step")
+        tokens_per_active = tokens_passed / self.args.get("model_n_active")
+        return {
+            f"{title}_tokens": {
+                "value": value,
+                "iteration": tokens_passed,
+            },
+            f"{title}_tokens_per_active": {
+                "value": value,
+                "iteration": tokens_per_active,
```
Comment (crewtool) [inline lizrd/support/logging.py:141]: ..active_parameters

File: lizrd/support/logging.py [hunk 11]
```diff
@@ -437,6 +485,8 @@ def get_logger(args, model, VOCAB_SIZE):
         logger_types = args.logger_types.split(",")
         assert len(logger_types) == len(set(logger_types)), "Duplicate logger types."
     initialized_loggers = []
+    add_logger_active_metrics(args)
```
Comment (crewtool) [inline lizrd/support/logging.py:488]: `get_logger` function should not change args

File: lizrd/support/logging.py [hunk 12]
```diff
@@ -120,12 +128,30 @@ def get_metric_with_flop_scale(self, value: float, iteration: int):
...
+        }
 
-        metric_x_flop = None
+    def get_auxiliary_metrics(
+        self, title: str, value: float, iteration: int, token_scale: bool = False
+    ):
         auxiliary_metrics = {}
+        if token_scale:
+            auxiliary_metrics = {
+                **auxiliary_metrics,
```
Comment (crewtool) [inline lizrd/support/logging.py:151]: auxiliary_metrics is empty

File: lizrd/support/logging.py (hunk 10 above)
Comment (ludziej) [reply lizrd/support/logging.py:141]: done

File: lizrd/support/logging.py (hunk 5 above)
Comment (ludziej) [reply lizrd/support/logging.py:224]: renamed to processed_token_scale

File: lizrd/support/misc.py (hunk 1 above)
Comment (ludziej) [reply lizrd/support/misc.py]: changed that
